RFC14: Add `per_resource` option to task's `count`

SteVwonder opened the pull request:
Adds `per_resource` option to a task's `count` key.

I noticed that at the top of the first section of examples (i.e., resource only requests), it states:
> In terms of a canonical jobspec, these requests are assumed to be requests to start an instance, i.e. run a single copy of flux start per allocated node.

But then use-cases 1.6 and 1.7 (1.6 and 1.8 in this PR), allocate and run `flux start` on a per-core basis.  Do we want to move those use-cases out of the `resource only requests` section of examples and into the 2nd section of examples?

Closes #150 
Closes #156

grondo commented:
> But then use-cases 1.6 and 1.7 (1.6 and 1.8 in this PR), allocate and run flux start on a per-core basis. Do we want to move those use-cases out of the resource only requests section of examples and into the 2nd section of examples?

Sorry for the delay, I thought I had already replied here.

We should think about whether the "resource only" requests description make much sense anymore. The examples launch brokers, so that term is perhaps confusing now?

 - UC1.6 runs one flux-broker per core, but there is also only one core per node. If the example is trying to demonstrate running a broker per core, it would be more enlightening perhaps to make the request for >1 core per node.

 - UC1.8 is similar -- 1 broker is launched per core only because that is the maximum amount of cores we could have available on a host.

I'm not sure there needs to be an example of running an instance with more than one broker per host/node, because that style should not be encouraged (at least I can't think of a reason). Instead we should maybe reword or rework the examples to show how starting an instance with at most 1 broker per node can be done with any resource request by use of your new `per_resource: { "type": "node", "count": 1 }` syntax.

SteVwonder commented:
>  Instead we should maybe reword or rework the examples to show how starting an instance with at most 1 broker per node can be done with any resource request by use of your new `per_resource: { "type": "node", "count": 1 }` syntax.

:+1: Agreed.  I dropped the new Use Case 1.7 which I added mainly to show a usage of `per_resource` and instead modified the pre-exising Use Cases 1.6 and 1.7 to only launch one broker per node and thus use `per_resource`.

> We should think about whether the "resource only" requests description make much sense anymore.

I changed the section titles to "Node-level Requests" and "General Requests".  I'm not in love with the latter name, but it works.  If someone has a better name (for either), I'm open to suggestions.

SteVwonder commented on spec_14.adoc:
```diff
@@ -221,20 +221,29 @@ descriptor SHALL contain the following keys:
- the keys `per_slot` and `total`, with other keys reserved for future
- or site-specific extensions.
-+
-[horizontal]
```
Just an FYI, I removed this `[horizontal]` because it made the `type` and `count` sub-list indent really far.  I thought it looked odd.

grondo commented:
> I changed the section titles to "Node-level Requests" and "General Requests". I'm not in love with the latter name, but it works. If someone has a better name (for either), I'm open to suggestions.

This works for me. Thanks!
